libraries/chain/wasm_interface_collection.cpp

```diff
@@ -9,26 +9,29 @@ namespace eosio::chain {
 
 #ifdef EOSIO_EOS_VM_OC_RUNTIME_ENABLED
 struct eosvmoc_tier {
+   // Called from main thread
    eosvmoc_tier(const std::filesystem::path& d, const eosvmoc::config& c, const chainbase::database& db)
       : cc(d, c, db) {
-      // construct exec for the main thread
-      init_thread_local_data();
+      // Construct exec and mem for the main thread
+      exec = std::make_unique<eosvmoc::executor>(cc);
+      mem  = std::make_unique<eosvmoc::memory>(wasm_constraints::maximum_linear_memory/wasm_constraints::wasm_page_size);
    }
 
-   // Support multi-threaded execution.
+   // Called from read-only threads
    void init_thread_local_data() {
       exec = std::make_unique<eosvmoc::executor>(cc);
+      mem  = std::make_unique<eosvmoc::memory>(eosvmoc::memory::sliced_pages_for_ro_thread);
    }
 
    eosvmoc::code_cache_async cc;
 
    // Each thread requires its own exec and mem.
    thread_local static std::unique_ptr<eosvmoc::executor> exec;
-   thread_local static eosvmoc::memory mem;
+   thread_local static std::unique_ptr<eosvmoc::memory>   mem;
```

Comment:
(obviously not in this PR) but I'd really like for us to noodle on how to move away from these `thread_local`s, they have dangerous lifetime semantics imo. In this particular example, when `eosvmoc_tier` is destroyed, `cc` is destroyed, but the `exec` created with a reference to that `cc` lives on. So if `executor`'s dtor (which will run well after `eosvmoc_tier`'s dtor) touches the `code_cache_async` that was passed to it via reference: bad news! It's a real trap -- passing a reference to some object to the ctor of an object implies that reference will be valid for the lifetime of the object.. but it's not true here.

Reply:
Thanks. I will be looking into moving away from those thread_locals.